tests/v3/test_store/test_stateful_store.py

```diff
@@ -0,0 +1,206 @@
+
+    async def store_clear(self):
+        await self.store.clear()
+
+    # async def listdir(self, group): #does listdir take group?
+
+    #    dir_ls = await self.store.list_dir(group)
+
+    #    return dir_ls
+    # ------
+
+    @rule(key=paths, data=st.binary(min_size=0, max_size=100))
+    def set(self, key: str, data: bytes) -> None:
+        note(f"(set) Setting {key!r} with {data}")
+        assert not self.store.mode.readonly
+        data_buf = Buffer.from_bytes(data)
+        asyncio.run(self.store_set(key, data_buf))
+        self.model[key] = data_buf  # this was data
+
+    @invariant()
+    def check_paths_equal(self) -> None:
+        note("Checking that paths are equal")
+        paths = asyncio.run(self.store_list())
+        # note(f'(check paths equal) {self.model=}, {self.store._store_dict=}')
+        assert list(self.model.keys()) == paths
+        assert len(self.model.keys()) == len(self.store._store_dict)
+        assert self.model == self.store._store_dict
+
+    @precondition(lambda self: len(self.model.keys()) > 0)
+    @rule(data=st.data())
+    # @rule(key=keys_bundle)
+    def get(self, data) -> None:
+        key = data.draw(st.sampled_from(sorted(self.model.keys())))
+        store_value = asyncio.run(self.get_key(key))
+        # note(f'(get) {self.model[key]=}, {store_value}')
+        # to bytes here necessary (on model and store) because data_buf set to model in set()
+        assert self.model[key].to_bytes() == store_value.to_bytes()
+
+    @precondition(lambda self: len(self.model.keys()) > 0)
+    @rule(data=st.data())
+    def get_partial_values(self, data) -> None:
+        key_st = st.sampled_from(sorted(self.model.keys()))
+        key_range = data.draw(key_ranges(keys=key_st))
+
+        obs_maybe = asyncio.run(self.get_partial(key_range))
+        observed = []
+
+        for obs in obs_maybe:
+            assert obs is not None
+            observed.append(obs.to_bytes())
+
+        model_vals_ls = []
+
+        for idx in range(len(observed)):
+            key, byte_range = key_range[idx]
+            model_vals = self.model[key]
+            start = byte_range[0] or 0
+            step = byte_range[1]
+            stop = start + step if step is not None else None
+            model_vals_partial = model_vals[start:stop]
+            model_vals_ls.append(model_vals_partial)
+
+        assert all(
+            obs == exp.to_bytes() for obs, exp in zip(observed, model_vals_ls, strict=True)
+        ), (
+            observed,
+            model_vals_ls,
+        )
+
+    @precondition(lambda self: len(self.model.keys()) > 0)
+    @rule(data=st.data())
+    def delete(self, data) -> None:
+        path_st = data.draw(st.sampled_from(sorted(self.model.keys())))
+        note(f"(delete) Deleting {path_st=}")
+
+        asyncio.run(self.delete_key(path_st))
+
+        del self.model[path_st]
+
+        # property tests
+        assert self.model.keys() == self.store._store_dict.keys()
```

Comment:
```suggestion
        assert self.model.keys() == self.store._store_dict.keys()
```

Don't you have a version of this in your invariant already? It's going to be run after every rule...